File: ChibuzorTasks/mydairy/src/mydairy/Dairy.java
package mydairy.src.mydairy;

import java.util.ArrayList;
import java.io.Serializable;

public class Dairy implements Serializable {
    private static final long serialVersionUID = 1L;
    private int id;
    private String name;
    private String pin;
    private ArrayList<DairyEntry> dairyEntries;
    private boolean isLocked;

    public Dairy(String name, String pin) {
        this.name = name;
        this.pin = pin;
        this.dairyEntries = new ArrayList<>();
        this.isLocked = false;
        this.id = 0;
    }

    public DairyEntry createDairyEntry(String title, String body) {
        if(this.isLocked) unlockDairy(pin);
        DairyEntry entry = new DairyEntry(++id, title, body);
        this.dairyEntries.add(entry);
        this.isLocked = true;
        return entry;
    }

    public void unlockDairy(String pin) {
        if(validatePin(pin)) this.isLocked = false;
    }

    public boolean validatePin(String pin) {
        if(pin.isEmpty()) throw new RuntimeException("Pin cannot be empty");
        if(!pin.equals(this.pin)) throw new RuntimeException("Invalid pin");
        return true;
    }

    public void lockDairy(String pin) {
        if(this.isLocked) throw new RuntimeException("Locked");
        if(validatePin(pin)) this.isLocked = true;
    }

    public boolean isLocked() {
        return this.isLocked;
    }

    public void deleteDairyEntryById(String pin, int id) {
        if(!validatePin(pin)) throw new RuntimeException("Invalid pin");
        if(findDairyEntryById(id)==null) throw new RuntimeException("Dairy entry not found");
        DairyEntry dairyEntry;
        dairyEntry = findDairyEntryById(id);
        this.dairyEntries.remove(dairyEntry);
    }

    public DairyEntry findDairyEntryById(int id) {
        if(dairyEntries.isEmpty()) return null;
        if(this.isLocked) throw new RuntimeException("Locked");
        for(DairyEntry entry : this.dairyEntries) {
            if(entry.getId() == id) {
                return entry;
            }
        }
        return null;
    }

    public int getDairyEntryCount() {
        return this.dairyEntries.size();
    }

    public String viewDairyEntry(String pin) {
        if(!validatePin(pin)) throw new RuntimeException("Invalid pin");
        this.unlockDairy(pin);
        StringBuilder entries = new StringBuilder();
        for(DairyEntry entry : this.dairyEntries) {
            entries.append("**************************");
            entries.append(entry.toString());
            entries.append("**************************");
            entries.append("\n");
        }
        return entries.toString();
    }

    public void updateEntryById(int id, String title, String body) {
        if(this.isLocked) throw new RuntimeException("Locked");
        DairyEntry dairyEntry = findDairyEntryById(id);
        dairyEntry.setTitle(title);
        dairyEntry.setBody(body);
    }

    public String getName() {
        return this.name;
    }
}

File: PythonClass/classWorks/Feb-2025/dairy/src/dairy.py
from dairy.src.dairy import *
from dairy.src.dairy.dairy_entry import DairyEntry


class Dairy:
    def __init__(self, name, pin):
        self.name = name
        self.pin = pin
        self.dairy_entries = []
        self.locked = True
        self.next_entry_id = 1

    def create_dairy_entry(self, title, body):
        if self.locked:
            raise RuntimeError("Dairy is locked. Unlock it first.")
        entry = DairyEntry(self.next_entry_id, title, body)
        self.dairy_entries.append(entry)
        self.next_entry_id += 1
        return entry

    def unlock_dairy(self, pin):
        if self.validate_pin(pin):
            self.locked = False

    def validate_pin(self, pin):
        if not pin:
            raise ValueError("PIN cannot be empty")
        if pin != self.pin:
            raise ValueError("Invalid PIN")
        return True

    def lock_dairy(self, pin):
        if self.validate_pin(pin):
            self.locked = True
        else:
            raise ValueError("Invalid PIN")

    def is_locked(self):
        return self.locked

    def delete_dairy_entry_by_id(self, pin, entry_id):
        if self.locked:
            raise RuntimeError("Dairy is locked. Unlock it first.")
        if not self.validate_pin(pin):
            raise ValueError("Invalid PIN")
        entry = self.find_dairy_entry_by_id(entry_id)
        if entry is None:
            raise ValueError("Dairy entry not found")
        self.dairy_entries.remove(entry)

    def find_dairy_entry_by_id(self, entry_id):
        if self.locked:
            raise RuntimeError("Dairy is locked. Unlock it first.")
        for entry in self.dairy_entries:
            if entry.get_entry_id() == entry_id:
                return entry
        return None

    def get_dairy_entry_count(self):
        return len(self.dairy_entries)

    def view_dairy_entries(self, pin):
        if not self.validate_pin(pin):
            raise ValueError("Invalid PIN")
        if self.locked:
            raise RuntimeError("Dairy is locked. Unlock it first.")
        return self.dairy_entries

    def update_entry_by_id(self, entry_id, title, body):
        if self.locked:
            raise RuntimeError("Dairy is locked. Unlock it first.")
        entry = self.find_dairy_entry_by_id(entry_id)
        if entry is None:
            raise ValueError("Dairy entry not found")
        entry.set_title(title)
        entry.set_body(body)

    def get_name(self):
        return self.name

    def __str__(self):
        return f"Dairy Name: {self.name}, Locked: {self.locked}, Entries: {self.get_dairy_entry_count()}"